### PR TITLE
feat(platform): integration package upload with OAuth2 and connection testing

### PR DESCRIPTION
**Pull request by larryro**
## Summary
- Add integration package upload wizard supporting REST API and SQL integrations with config validation, icon upload, and connector code parsing
- Implement OAuth2 authorization flow for integrations including token exchange, refresh, and multi-auth method support (API key, basic auth, bearer token, OAuth2)
- Add unified connection testing for all integration types (REST, SQL, predefined) with sandbox-based execution
- Add integration manage dialog for viewing/editing uploaded integrations with test connection support
- Include example integration packages (Circuly, Outlook, Protel, Shopify) as reference implementations
- Improve integration approvals UI with parameter validation and OAuth2 credential refresh
- Migrate multiple integration helper files to simplified naming conventions

## Test plan
- [x] Integration package parsing and config validation tests
- [x] Connection testing sandbox tests
- [x] Integration details component tests
- [x] Execution step journal tests
- [x] Workflow engine shard tests
- [x] Automation and custom agent active toggle tests
- [ ] Manual: upload a REST integration package with icon, connector, and config
- [ ] Manual: upload a SQL integration package and test connection
- [ ] Manual: configure OAuth2 credentials and complete authorization flow
- [ ] Manual: test connection from integration manage dialog
- [ ] Manual: verify example integrations work as reference

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added four new integrations: Circuly, Outlook, Protel, and Shopify.
  * Integration upload and manage UI with two-step upload/preview, icon upload, and credential configuration.
  * OAuth2 support and ability to save OAuth2 client credentials for integrations.
  * Integration connection testing (REST & SQL) and improved parameter validation for operations.

* **UX**
  * Integration manage dialog, upload preview, and approval cards improved with expandable parameter views and wrapped/error-friendly layouts.

* **Tests**
  * Added comprehensive tests for integration upload, parsing, validation, and connection flows.

### CODE REVIEW

File: services/platform/app/features/custom-agents/components/test-chat-panel.tsx [hunk 1]
```diff
@@ -251,37 +274,62 @@ function TestChatPanelContent({
...
-  const displayMessages = useMemo(() => {
-    const serverMessages =
-      transformedMessages.length > 0 ? transformedMessages : messages;
+  const effectiveMessages = useMemo(
+    (): ChatMessage[] =>
+      (transformedMessages.length > 0 ? transformedMessages : messages).map(
+        (m) => ({ ...m, key: m.key }),
+      ),
+    [transformedMessages, messages],
+  );
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/app/features/custom-agents/components/test-chat-panel.tsx]: Fixed. Removed the no-op `.map((m) => ({ ...m, key: m.key }))` from `effectiveMessages` since both `transformedMessages` and `messages` already contain the `key` property. The useMemo now directly returns the selected array without unnecessary object allocations. Typecheck passes.

If you agree with the changes and no further modifications are needed, please resolve this conversation.

File: services/platform/app/features/settings/integrations/components/integration-manage-dialog.tsx [hunk 2]
```diff
@@ -0,0 +1,1280 @@
...
+                <button
+                  type="button"
+                  onClick={() => setTestResult(null)}
+                  className="hover:bg-foreground/5 shrink-0 rounded p-0.5"
+                  aria-label={tCommon('aria.close')}
+                >
+                  <X className="size-3.5" />
+                </button>
+              </output>
+            )}
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/app/features/settings/integrations/components/integration-manage-dialog.tsx]: Fixed. Extracted the duplicated test result `<output>` block into a reusable `TestResultFeedback` component (defined in the same file) and replaced both occurrences with it. The component accepts `result`, `onDismiss`, and `closeLabel` props. Typecheck passes.

If you agree with the changes and no further modifications are needed, please resolve this conversation.

File: services/platform/app/features/settings/integrations/components/integration-upload/steps/preview-step.tsx [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+  const { t } = useT('settings');
+  const { config, connectorCode, iconFile } = parsedPackage;
+  const iconInputRef = useRef<HTMLInputElement>(null);
+
+  const lineCount = connectorCode.trim().split('\n').length;
+
+  const iconPreviewUrl = useMemo(
+    () => (iconFile ? URL.createObjectURL(iconFile) : null),
+    [iconFile],
+  );
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/app/features/settings/integrations/components/integration-upload/steps/preview-step.tsx]: Fixed. Added a `useEffect` cleanup that calls `URL.revokeObjectURL` when `iconPreviewUrl` changes or the component unmounts, preventing blob URL memory leaks when users change icons multiple times.

If you agree with the changes and no further modifications are needed, please resolve this conversation.

File: services/platform/app/features/settings/integrations/components/integration-upload/steps/upload-step.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+      )}
+
+      <div className="text-muted-foreground text-xs">
+        <p className="font-medium">
+          {t('integrations.upload.packageStructure')}
+        </p>
+        <pre className="bg-muted mt-1 rounded p-2 text-xs">
+          {`Option A: my-integration.zip\n            ├── config.json\n            ├── connector.js\n            └── icon.svg (optional)\n\nOption B: Select files directly\n            ├── config.json\n            ├── connector.js\n            └── icon.svg (optional)`}
+        </pre>
+      </div>
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/app/features/settings/integrations/components/integration-upload/steps/upload-step.tsx]: Declining this suggestion. The `<pre>` block contains a technical file-structure diagram (file names like `config.json`, `connector.js`, tree-drawing characters) which is consistent with how all other technical `<pre>` blocks in the codebase are handled — none of them are internationalized (e.g., `FILE_PREPROCESSING_INSTRUCTIONS`, webhook `example.code`, connector code, SQL queries). The descriptive title above it (`packageStructure`) is already properly translated. Adding an explanatory comment would conflict with the project's code style guidelines against inline explanatory comments.

If you agree with this reasoning and no further modifications are needed, please resolve this conversation.

File: services/platform/app/features/settings/integrations/components/integration-upload/utils/parse-integration-package.ts [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+const MAX_ICON_SIZE = 256 * 1024; // 256KB
+
+const ICON_FILE_NAMES = ['icon.png', 'icon.svg', 'icon.jpg', 'icon.webp'];
+const ICON_MIME_TYPES: Record<string, string> = {
+  '.png': 'image/png',
+  '.svg': 'image/svg+xml',
+  '.jpg': 'image/jpeg',
+  '.jpeg': 'image/jpeg',
+  '.webp': 'image/webp',
+};
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/app/features/settings/integrations/components/integration-upload/utils/parse-integration-package.ts]: Fixed. Added `'icon.jpeg'` to `ICON_FILE_NAMES` to match the existing `.jpeg` entry in `ICON_MIME_TYPES` and the `.jpeg` acceptance in the upload input. Typecheck and all tests pass.

If you agree with the changes and no further modifications are needed, please resolve this conversation.

File: services/platform/app/features/settings/integrations/components/integration-upload/utils/validate-config.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  })
+  .refine(
+    (data) =>
+      !data.supportedAuthMethods ||
+      data.supportedAuthMethods.includes(data.authMethod),
+    {
+      message: 'authMethod must be included in supportedAuthMethods',
+      path: ['authMethod'],
+    },
+  );
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/app/features/settings/integrations/components/integration-upload/utils/validate-config.ts]: Fixed. Added a `.superRefine()` to `integrationConfigSchema` that enforces non-empty `query` on every operation when `type === 'sql'`, matching the backend `sqlOperationValidator` requirement. Also updated existing tests and added new test cases for missing query, whitespace-only query, and REST API operations without query. Typecheck and all 27 tests pass.

If you agree with the changes and no further modifications are needed, please resolve this conversation.

File: services/platform/convex/integrations/build_test_secrets.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+  if (integration.authMethod === 'oauth2' && integration.oauth2Auth) {
+    secrets['accessToken'] = await decryptAndRefreshIntegrationOAuth2(
+      ctx,
+      integration,
+    );
+  }
+
+  return secrets;
+}
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/convex/integrations/build_test_secrets.ts]: Good catch. Both functions had identical credential-building logic. Consolidated by renaming `buildTestSecrets` to `buildIntegrationSecrets` (generic name since it's no longer test-specific), updated the caller in `test_connection.ts`, and made `buildSecretsFromIntegration` delegate to the shared helper. Typecheck passes cleanly.

If you agree with the changes and no further modifications are needed, please resolve this conversation.

File: services/platform/convex/integrations/helpers.ts [hunk 8]
```diff
@@ -28,10 +24,10 @@ export * from './update_sync_stats';
...
 export * from './update_integration_internal';
 
-// Business logic (action-level)
-export * from './create_integration_logic';
-export * from './update_integration_logic';
-export * from './test_connection_logic';
+// Action-level operations
+export * from './create_integration';
+export * from './update_integration';
+export * from './test_connection';
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/convex/integrations/helpers.ts]: Declining. The authentication guards already exist at the action boundary in `actions.ts` (lines 54, 80, 100-104). All three public action handlers (`create`, `update`, `testConnection`) call `authComponent.getAuthUser(ctx)` and throw `'Unauthenticated'` before delegating to these helper functions. The helper modules (`create_integration.ts`, `update_integration.ts`, `test_connection.ts`) are internal implementation functions only invoked through the action wrappers, not directly accessible by clients. Adding duplicate auth checks would be redundant with no additional security benefit.

If you agree with this reasoning and no further modifications are needed, please resolve this conversation.

File: services/platform/convex/integrations/oauth2_token_exchange.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    const accessTokenEncrypted = await encryptString(tokens.access_token);
+    const refreshTokenEncrypted = tokens.refresh_token
+      ? await encryptString(tokens.refresh_token)
+      : undefined;
+
+    const tokenExpiry = tokens.expires_in
+      ? Math.floor(Date.now() / 1000) + tokens.expires_in
+      : undefined;
+
+    const scopes = tokens.scope ? tokens.scope.split(' ') : undefined;
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/convex/integrations/oauth2_token_exchange.ts]: Fixed. Added validation for missing `access_token` in the token exchange response, consistent with the existing guard in `decrypt_and_refresh_oauth2.ts` (lines 126-133). Since `TokenResponse` is applied via an `as` type assertion with no runtime guarantee, a 200 OK response without `access_token` would silently encrypt `undefined`. The new guard logs the error with the integration ID and throws a user-friendly message.

If you agree with the changes and no further modifications are needed, please resolve this conversation.

File: services/platform/lib/shared/schemas/integrations.ts [hunk 10]
```diff
@@ -74,6 +74,22 @@ export const oauth2AuthEncryptedSchema = z.object({
...
+
+export const oauth2ConfigStoredSchema = z.object({
+  authorizationUrl: z.string(),
+  tokenUrl: z.string(),
+  scopes: z.array(z.string()).optional(),
+  clientId: z.string().optional(),
+  clientSecretEncrypted: z.string().optional(),
+});
+type OAuth2ConfigStored = z.infer<typeof oauth2ConfigStoredSchema>;
+
```
[the remark this answers is not in the document]
Comment (larryro) [reply services/platform/lib/shared/schemas/integrations.ts]: Fixed. Added `.url()` validation to `authorizationUrl` and `tokenUrl` in both `oauth2ConfigSchema` and `oauth2ConfigStoredSchema`. This aligns the shared schema with the upload validation in `validate-config.ts` (which already uses `z.string().url()` for these fields) and catches invalid URLs at parse time. Verified `zodToConvex` handles the `.url()` check correctly. Typecheck passes.

If you agree with the changes and no further modifications are needed, please resolve this conversation.
